### PR TITLE
NED Refactoring (4)

### PR DESCRIPTION
**Pull request by Mehrad0711**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -149,7 +149,7 @@ It will also generate several files in <results_temp> folder. In `eval_bootleg/[
 
 Now you can use the extracted features from bootleg in downstream tasks such as semantic parsing to improve named entity understanding and consequently generation:
 ```bash
-genienlp train --train_tasks <train_task_names> --train_iterations <iterations> --preserve_case --save <savedir> --data <dataset_dir> --model TransformerSeq2Seq --pretrained_model facebook/bart-base --train_batch_tokens 1000 --val_batch_size 1000 --do_ned --database_dir <database_dir> --ned_retrieve_method bootleg --entity_attributes type_id type_prob --add_entities_to_text append --bootleg_model <bootleg_model>
+genienlp train --train_tasks <train_task_names> --train_iterations <iterations> --preserve_case --save <savedir> --data <dataset_dir> --model TransformerSeq2Seq --pretrained_model facebook/bart-base --train_batch_tokens 1000 --val_batch_size 1000 --do_ned --database_dir <database_dir> --ned_class_name bootleg --entity_attributes type_id type_prob --add_entities_to_text append --bootleg_model <bootleg_model>
```
Comment (gcampax) [inline README.md]: Please undo this change, ned_retrieve_method is a much more informative argument.

File: genienlp/ned/ned_utils.py [hunk 2]
```diff
@@ -281,18 +281,18 @@ def init_ned_model(args, ned_class_name=None):
     ned_model = None
     if args.do_ned:
         if ned_class_name is None:
-            if args.ned_retrieve_method == 'bootleg':
+            if args.ned_class_name == 'bootleg':
                 ned_class_name = 'Bootleg'
-            elif args.ned_retrieve_method == 'bootleg-annotator':
+            elif args.ned_class_name == 'bootleg-annotator':
```
Comment (gcampax) [inline genienlp/ned/ned_utils.py]: ned_retrieve_method can never be bootleg-annotator, right? The difference between bootleg and bootleg-annotator is whether you're called from "server" or "train" / "predict"

File: genienlp/ned/base.py [hunk 3]
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2020-2021 The Board of Trustees of the Leland Stanford Junior University
+# Copyright (c) 2021-2022 The Board of Trustees of the Leland Stanford Junior University
```
Comment (gcampax) [inline genienlp/ned/base.py]: 2022?

File: genienlp/ned/base.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        self.unk_id = 0
+        self.unk_type = self.id2typeqid[self.unk_id]
+        self.ckpt_name = None
+
+    def update_wiki2normalized_type(self):
+        for normalized_type, titles in self.almond_type_mapping.items():
+            for title in titles:
+                self.wiki2normalized_type.append((title, normalized_type))
+
+    def process_examples(self, examples, split_path, utterance_field):
```
Comment (gcampax) [inline genienlp/ned/base.py]: How can this method have an implementation if this is an abstract class?

File: genienlp/ned/base.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+                continue
+            ent_num_tokens = len(ent.split(' '))
+            idx = tokens_text.index(ent)
+            token_pos = len(tokens_text[:idx].strip().split(' '))
+            type = self.typeqid2id.get(self.alias2type[ent], self.unk_id)
+            tokens_type_ids[token_pos : token_pos + ent_num_tokens] = [[type] * self.max_features_size] * ent_num_tokens
+
+        return tokens_type_ids
+
+    def lookup(self, tokens, database_lookup_method=None, min_entity_len=2, max_entity_len=4):
```
Comment (gcampax) [inline genienlp/ned/base.py]: This method is not used by Bootleg right? So it doesn't belong here. It should go into a subclass.

File: genienlp/ned/bootleg.py [hunk 6]
```diff
@@ -30,80 +30,44 @@
...
-        self.cur_entity_embed_size = 0
-
-        # Mapping between model directory and checkpoint name
-        model2checkpoint = {'bootleg_uncased_mini': 'bootleg_wiki.pth', 'bootleg_uncased_super_mini': 'bootleg_wiki.pth'}
-
-        self.ckpt_name, extension = model2checkpoint[self.args.bootleg_model].split('.')
-        self.model_ckpt_path = os.path.join(self.model_dir, self.ckpt_name + '.' + extension)
+        ###
+
+        self.ckpt_name = 'bootleg_wiki'
```
Comment (gcampax) [inline genienlp/ned/bootleg.py:69]: This is a cmdline argument right?

File: genienlp/ned/bootleg.py [hunk 7]
```diff
@@ -173,21 +132,16 @@ def extract_mentions(self, input_path):
...
-        if len(tokens) > max_size:
-            tokens = tokens[:max_size]
-        else:
-            tokens += [pad_id] * (max_size - len(tokens))
-        return tokens
-
     def disambiguate_mentions(self, config_args):
         run_model(self.args.bootleg_dump_mode, config_args)
 
     def post_process_bootleg_types(self, title):
```
Comment (gcampax) [inline genienlp/ned/bootleg.py:138]: Why is the initialization for type postprocessing done in the parent class, but type postprocessing is only used in Bootleg?

File: genienlp/ned/bootleg.py [hunk 8]
```diff
@@ -323,103 +215,70 @@ def collect_features_per_line(self, line, threshold):
...
+        self.create_jsonl(path, examples, utterance_field)
+
+        # extract mentions and mention spans in the sentence and write them to output jsonl files
+        self.extract_mentions(path)
+
+        # find the right entity candidate for each mention
+        self.disambiguate_mentions(config_args)
+
+
+class ServingBootlegEntityDisambiguator(BatchBootlegEntityDisambiguator):
```
Comment (gcampax) [inline genienlp/ned/bootleg.py:279]: Do you need this inheritance? What are you inheriting?

File: genienlp/ned/main.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+import marisa_trie
+import ujson
+
+from ..data_utils.almond_utils import quoted_pattern_with_space
+from .base import AbstractEntityDisambiguator
+
+logger = logging.getLogger(__name__)
+
+
+class NaiveEntityDisambiguator(AbstractEntityDisambiguator):
```
Comment (gcampax) [inline genienlp/ned/main.py]: How can the classes in this file not implement process_examples?

File: genienlp/ned/main.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    def __init__(self, args):
+        super().__init__(args)
+
+        with open(os.path.join(self.args.database_dir, 'es_material/alias2type.json'), 'r') as fin:
+            # alias2type.json is a big file (>4G); load it only when necessary
```
Comment (gcampax) [inline genienlp/ned/main.py:213]: Why is alias2type a JSON file at all, and not a binary-serialized marisa trie?

**Review by gcampax — CHANGES_REQUESTED** [review]
This is definitely a good cleanup, and we're getting there, but we can still do better.

File: genienlp/ned/base.py (hunk 5 above)
Comment (Mehrad0711) [reply genienlp/ned/base.py]: Ok yeah I should create a base class for Naive and Oracle and many of these methods should move there.

File: genienlp/ned/base.py (hunk 4 above)
Comment (Mehrad0711) [reply genienlp/ned/base.py]: moved to base class

File: genienlp/ned/main.py (hunk 9 above)
Comment (Mehrad0711) [reply genienlp/ned/main.py:43]: will get the method from abstract

File: genienlp/ned/ned_utils.py (hunk 2 above)
Comment (Mehrad0711) [reply genienlp/ned/ned_utils.py]: yes. that's why there's the option of passing `ned_class_name` as an argument to override what args say.
Alternatively, we can pass a `mode` which can be (server|batch) but I avoided that since server mode and batch mode for non bootleg classes is the same.

File: genienlp/ned/main.py (hunk 10 above)
Comment (Mehrad0711) [reply genienlp/ned/main.py:213]: Good point. I will convert it.

File: genienlp/ned/bootleg.py (hunk 6 above)
Comment (Mehrad0711) [reply genienlp/ned/bootleg.py:69]: Checkpoint name used to be different for different bootleg models. Now it's all the same. We still can pass different bootleg_models.

File: genienlp/ned/bootleg.py (hunk 8 above)
Comment (Mehrad0711) [reply genienlp/ned/bootleg.py:279]: We discussed this on slack. `ServingBootlegEntityDisambiguator` uses several methods and attributes from `BatchBootlegEntityDisambiguator`. Turning them into utility functions is not viable since methods are called within each other.

File: genienlp/ned/bootleg.py (hunk 7 above)
Comment (Mehrad0711) [reply genienlp/ned/bootleg.py:138]: Oracle uses `wiki2normalized_type` to normalize types too. Except only uses the first half of this function.  I have defined a base `normalize_types` method that other classes would inherit in a new PR I'm working on. That would solve this confusion.

**Review by gcampax — APPROVED** [review]
Alright, I think this is a very nice cleanup, let's get it in!
